pkg/codegen/dotnet/gen_program_expressions.go

```diff
@@ -0,0 +1,555 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dotnet
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/schema"
+	"io"
+	"math/big"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2/model"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"
+	"github.com/zclconf/go-cty/cty"
+)
+
+type nameInfo int
+
+func (nameInfo) Format(name string) string {
+	return validIdentifier(name)
+}
+
+// rewriteExpression amends the expression with intrinsics for C# generation.
+func (g *generator) rewriteExpression(expr model.Expression, typ model.Type) model.Expression {
+	expr, diags := hcl2.RewriteApplies(expr, nameInfo(0), !g.asyncInit)
+	contract.Assert(len(diags) == 0)
+	expr = hcl2.RewriteConversions(expr, typ)
+	if g.asyncInit {
+		expr = g.awaitInvokes(expr)
+	} else {
+		expr = g.outputInvokes(expr)
+	}
+	return expr
+}
+
+// outputInvokes wraps each call to `invoke` with a call to the `output` intrinsic. This rewrite should only be used if
+// resources are instantiated within a stack constructor, where `await` operator is not available. We want to avoid the
+// nastiness of working with raw `Task` and wrap it into Pulumi's Output immediately to be able to `Apply` on it.
```

Comment:
Could we use leave these as `Task<>`s and use [`ContinueWith<>`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task-1.continuewith?view=netcore-3.1#System_Threading_Tasks_Task_1_ContinueWith__1_System_Func_System_Threading_Tasks_Task__0____0__) here instead? That seems more idiomatic.

Reply:
FTW, all our examples (in the examples repo) convert to outputs immediately. IMO, `ContinueWith` is quite crappy and most people aren't familiar with it either, so I'd rather encourage them to learn outputs instead.

Reply:
Let's add this comment from `awaitInvokes` to this as well, then:

```
// Note that this depends on the fact that invokes are the only way to introduce promises
// in to a Pulumi program; if this changes in the future, this transform will need to be applied in a more general way
// (e.g. by the apply rewriter).
```